### ISSUE TITLE
Sequential subject selection advancing the subject by 10 when requesting subjects

### ISSUE THREAD

**Issue by eatyourgreens** [zooniverse/panoptes#3789]
An odd bug when using Designator with ordered (prioritised) workflows.

Open this link in three different tabs:
https://frontend.preview.zooniverse.org/projects/eatyourgreens/-project-testing-ground/classify/workflow/3586/subject-set/4941?env=staging&demo=true

What I'm seeing is the following:
- first tab: subject 1.
- second tab: subject 11.
- third tab: subject 21.

Designator is remembering which subjects it has sent me, then advancing the queue by 10 in each new tab.

You can also reproduce this bug, in a single tab, by changing to a new subject set then back to the original set. On returning, the queue has advanced by 10, even though I haven't classified anything.

Expected behaviour: I think I should always get subject 1, at least until I submit a classification for it.

**Comment by camallen** [comment]
This isn't a bug but rather a feature - we keep a local cache of what Designator has recently handed out to a user. This cache of `subject_ids` is used to filter subsequent requests for subjects  to avoid folks getting repeat subjects when the designator system and the API get out of sync about what a user has classified. 

Also note - this only impacts logged in users
- https://designator-staging.zooniverse.org/api/workflows/3586?subject_set_id=4931?user_id=6

Non logged in will return the ordering correctly as we don't cache / filter the data as we don't know what they have seen.
- https://designator-staging.zooniverse.org/api/workflows/3586?subject_set_id=4931

If we turn this functionality off on priority workflows we run the risk of returning duplicates during periods when the API is busy and we get some state drift between designator and panoptes API. Perhaps the client side can track this data and filter on that end? 

As long as folks are ok with possible impact then we can implement a skip of this filter if a workflow is configured for priority selection. A possible implementation would be to 
1. Pass the `workflow.priority` instance to the filtering function here, https://github.com/zooniverse/designator/blob/986401cc2538030f36f62d81716d9e1a2d1b15c6/lib/designator/selection.ex#L30
2. Add two methods (using Elixir pattern matching) for the boolean priority values https://github.com/zooniverse/designator/blob/986401cc2538030f36f62d81716d9e1a2d1b15c6/lib/designator/selection.ex#L126-L128
    - `reject_recently_selected(stream, user, true)` -> return the stream without filtering
    - `reject_recently_selected(stream, user, false)` -> do the current filtering implementation

**Comment by eatyourgreens** [comment]
>If we turn this functionality off on priority workflows we run the risk of returning duplicates during periods when the API is busy and we get some state drift between designator and panoptes API. Perhaps the client side can track this data and filter on that end?

You're asking for the client to persist state across tabs (or windows), or even page refreshes. That's simply not possible. Refreshing the page, or opening the same URL in different tabs, should always show the same subject for ordered sets. Refreshing the page should not increment the current position by 10.

**Comment by eatyourgreens** [comment]
For what it's worth, Designator is sending duplicates to logged-in volunteers on HMS NHS, and has been doing this for some time. That's what originally prompted me to look into subject selection:
https://www.zooniverse.org/projects/msalmon/hms-nhs-the-nautical-health-service/talk/4936/2331577?comment=3847847

What's happening there is that the task is quite long, so requests to `/subjects/queued` are more than five minutes apart. I wrote up the bug here:
https://github.com/zooniverse/front-end-monorepo/issues/2818

I think the conclusion from that is that a five minute cache isn't useful on a real project, anyway.

**Comment by eatyourgreens** [comment]
By the way, volunteers have been working around those duplicates by going to the home page, changing workflow to clear the subject queue, then changing back to the original workflow, in order to trigger a fresh request to Designator. This post has a very detailed walkthrough:
https://www.zooniverse.org/projects/msalmon/hms-nhs-the-nautical-health-service/talk/4936/2331577?comment=3854001

If I'm understanding the cache correctly, the problem with that approach is that they could then skip forward 10 subjects, which leaves unclassified subjects in the chosen subject set. Since they're classifying the set in order, they will never be shown those skipped subjects.

**Comment by camallen** [comment]
> You're asking for the client to persist state across tabs (or windows), or even page refreshes. That's simply not possible. Refreshing the page, or opening the same URL in different tabs, should always show the same subject for ordered sets. 

No not across tabs etc just in JS memory like we do for PFE etc https://github.com/zooniverse/Panoptes-Front-End/blob/45399f3267f45fa1d67d5bcd162b4050ed11e5ac/app/redux/ducks/classify.js#L355

Agree we don't want to look into state management across tabs, i may be wrong but on most projects, most volunteers don't open multiple tabs with the classifier open (maybe they have talk up?). Perhaps transcription projects are rather different and we need to support that.

> EDIT: we could persist state between client and server by passing an optional subject_id as a parameter in the API request ie. give me 10 unclassified subjects, in order, starting from the specified ID. Or priority if that's easier to use. It won't be known on first entering the page, but subsequent requests will always want to start from the last subject in the current queue.

That's interesting - providing the known subject id to select beyond could be very useful as a seed for the starting point in the available subjects stream.

> That's what originally prompted me to look into subject selection:
https://www.zooniverse.org/projects/msalmon/hms-nhs-the-nautical-health-service/talk/4936/2331577?comment=3847847
> 
> What's happening there is that the task is quite long, so requests to /subjects/queued are more than five minutes apart. I wrote up the bug here:
https://github.com/zooniverse/front-end-monorepo/issues/2818

Seems, like this is triggered by the priority setting and if designator's recently selected cache is expired, thus selecting before reaching the end of the current subject queue will return some ids currently in the queue as the API / Designator don't know that it's been seen before (thus those subject ids are in the available pool). 

Seems you've fixed this in https://github.com/zooniverse/front-end-monorepo/pull/2819

> I think the conclusion from that is that a five minute cache isn't useful on a real project, anyway.

It is for the majority of Zooniverse projects, especially ones that have fairly short workflow completion times. I agree it's probably not very useful in long running transcription task projects and especially not for workflows that rely on the `priority` setting.


> If I'm understanding the cache correctly, the problem with that approach is that they could then skip forward 10 subjects, which leaves unclassified subjects in the chosen subject set. Since they're classifying the set in order, they will never be shown those skipped subjects.

Yep - this is correct. I think we agree the need to modify the designator behaviour for workflows that use the `priority` setting. We can skip / disable the recently seen filtering for workflows that have this setting.

**Comment by eatyourgreens** [comment]
>No not across tabs etc just in JS memory like we do for PFE etc https://github.com/zooniverse/Panoptes-Front-End/blob/45399f3267f45fa1d67d5bcd162b4050ed11e5ac/app/redux/ducks/classify.js#L355
>
>Agree we don't want to look into state management across tabs, i may be wrong but on most projects, most volunteers don't open multiple tabs with the classifier open (maybe they have talk up?). Perhaps transcription projects are rather different and we need to support that.

I used multiple tabs to easily illustrate the bug, but you'd also run into it on page refresh in the same tab. Remember that JS will be unloaded and cleared when the page unloads, during a refresh.

We're experimenting with persisting the subject queue in session storage, but that has lead to other problems for volunteers, where session storage persists a subject that they have classified, and we're back to them being stuck on a page that they can't get past. A workaround for that is to shut down the active tab and re-open the site in a new tab, but then that runs into the danger of skipping forward by 10 pages.

Another bug we've seen, which has been occupying me for the last month or so, is duplicate subjects after returning from Talk. Volunteers report getting duplicates from Designator after pressing Done & Talk, then coming back to the Classify page only to load the subject that they just submitted from Designator. I've been able to reproduce it (so has @mrniaboc and @snblickhan). Could that happen because they're coming back from Talk before the user cache has updated? I've been struggling to understand why Designator would serve up a subject that I've just classified.

The original bug report for that is here:
https://github.com/zooniverse/front-end-monorepo/issues/2610
